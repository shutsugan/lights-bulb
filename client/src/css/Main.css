.Main {
	width: 100%;
	height: auto;
	display: flex;
	flex-flow: row nowrap;
	justify-content: center;
	padding: 16px;
	position: relative;
}

@media only screen and (max-width: 860px) {
    .Main {
		flex-flow: column-reverse;
		min-width: 430px;
	}
}