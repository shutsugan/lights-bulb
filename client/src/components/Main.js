import React, {Component} from 'react';
import MainTable from './MainTable';
import Slider from './Slider';

import '../css/Main.css';

class Main extends Component {
	constructor() {
		super();

		this.state = {
			current: {},
			devices: [],
		}
	}

	componentDidMount() {
		this._fetch();
	}

	/**
 	 * Get the data from the api.
	 */
	_fetch = _ => {
		fetch('/api/v1/device')
			.then(res => res.json())
			.then(res => this.setState({devices: res.data}));
	}

	/**
 	 * Get the current row.
 	 * @param {object} current the current row to be interacted with
	 */
	_getCurrent = current => {
		this.setState({current: current});
	}

	/**
 	 * Update the light bubl.
 	 * @param {number} id the id of the light bulb
 	 * @param {object} data the data of the updated light bulb
	 */
	_update = (id, data) => {
		const options = {
			method: 'PATCH',
			headers: {'content-type': 'application/json'},
			body: JSON.stringify({data})
		};

		fetch(`/api/v1/device/${id}`, options)
			.then(res => res.json())
			.then(res => {
				const new_device = res.data;
				const devices = this.state.devices.map(device => {
					if (device.id === new_device.id) return new_device;
					else return device;
				});

				this.setState({
					devices: devices,
					current: new_device
				});
			});
	}

	render() {
		return(
			<div className="Main">
				<MainTable
					devices={this.state.devices} 
					current={this.state.current} 
					_update={this._update} 
					_getCurrent={this._getCurrent} 
				/>
				<Slider 
					current={this.state.current} 
					_update={this._update} 
					_getCurrent={this._getCurrent}
				/>
			</div>
		);
	}
}

export default Main;