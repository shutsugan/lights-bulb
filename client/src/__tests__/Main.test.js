import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import Main from '../components/Main';

describe('Main', () => {
    it('should be defined', () => {
      expect(Main).toBeDefined();
    });
  
    it('should render correctly', () => {
      const tree = shallow(<Main />);
      expect(tree).toMatchSnapshot();
    });
});