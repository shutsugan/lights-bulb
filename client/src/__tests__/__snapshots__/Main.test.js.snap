// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="Main"
>
    <MainTable
        _getCurrent={[Function]}
        _update={[Function]}
        current={Object {}}
        devices={Array []}
    />
    <Slider
        _getCurrent={[Function]}
        _update={[Function]}
        current={Object {}}
    />
</div>,
  "nodes": Array [
    <div
      className="Main"
>
      <MainTable
            _getCurrent={[Function]}
            _update={[Function]}
            current={Object {}}
            devices={Array []}
      />
      <Slider
            _getCurrent={[Function]}
            _update={[Function]}
            current={Object {}}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_context": Object {},
    "_element": <Main />,
    "_instance": Main {
      "_fetch": [Function],
      "_getCurrent": [Function],
      "_update": [Function],
      "context": Object {},
      "props": Object {},
      "refs": Object {},
      "state": Object {
        "current": Object {},
        "devices": Array [],
      },
      "updater": Updater {
        "_renderer": [Circular],
      },
    },
    "_newState": null,
    "_rendered": <div
      className="Main"
>
      <MainTable
            _getCurrent={[Function]}
            _update={[Function]}
            current={Object {}}
            devices={Array []}
      />
      <Slider
            _getCurrent={[Function]}
            _update={[Function]}
            current={Object {}}
      />
</div>,
    "_rendering": false,
    "_updater": Updater {
      "_renderer": [Circular],
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Main />,
}
`;
